controllers/clustertopology_controller_compute.go

```diff
@@ -140,9 +141,129 @@ func (r *ClusterTopologyReconciler) getClass(ctx context.Context, cluster *clust
+		infra, err = r.getCurrentInfrastructureClusterState(ctx, cluster)
+		if err != nil {
+			return nil, errors.Wrap(err, "current cluster state not found")
+		}
+	}
+	// Reference to the control plane can be nil, and is expected to be on the first reconcile. In this case the method
+	// should still be allowed to continue
+	if cluster.Spec.ControlPlaneRef != nil {
+		cp, err = r.getControlPlaneState(ctx, cluster, class)
+		if err != nil {
+			return nil, errors.Wrap(err, "current cluster state not found")
+		}
+	}
+	// A cluster may have zero or more machine deployments and a cluster is expected to have zero machine deployments on
+	// first reconcile
+	m, err = r.getCurrentMachineDeploymentState(ctx, cluster)
+	if err != nil {
+		return nil, errors.Wrap(err, "current cluster state not found")
+	}
+	return &clusterTopologyState{
+		cluster:               cluster,
+		infrastructureCluster: infra,
+		machineDeployments:    m,
+		controlPlane:          cp,
+	}, nil
+}
+
+// getCurrentInfrastructureClusterState looks for the infrastructure underlying the cluster. If a reference is set but not
+// found, through an error or not existing in the api server an error is thrown.
+func (r *ClusterTopologyReconciler) getCurrentInfrastructureClusterState(ctx context.Context, cluster *clusterv1.Cluster) (*unstructured.Unstructured, error) {
+	infra, err := r.getReference(ctx, cluster.Spec.InfrastructureRef)
+	if err != nil {
+		return nil, errors.Wrap(err, "cluster infrastructure state")
+	}
+	return infra, nil
+}
+
+// getControlPlaneState returns information on the control plane being used by the cluster; If a reference not found,
+// an error is thrown. Also, if the control plane requires infrastructure machines as configured in the clusterClass
+// an error will be thrown if the control plane is found but with no infrastructure machine templates.
+func (r *ClusterTopologyReconciler) getControlPlaneState(ctx context.Context, cluster *clusterv1.Cluster, class clusterv1.ClusterClass) (controlPlaneTopologyState, error) {
+	cp, err := r.getReference(ctx, cluster.Spec.ControlPlaneRef)
+	if err != nil {
+		return controlPlaneTopologyState{}, errors.Wrap(err, "control plane")
+	}
+	// Some control plane providers may not require infrastructure machines to run. This check returns early if the field
+	// indicating infrastructure is required is not found in the clusterClass of the given cluster
+	if class.Spec.ControlPlane.MachineInfrastructure == nil {
+		return controlPlaneTopologyState{
+			cp,
+			nil,
+		}, nil
+	}
+	cpInfraRef, err := getNestedRef(cp, "spec", "machineTemplate", "infrastructureRef")
+	if err != nil {
+		return controlPlaneTopologyState{cp, nil}, errors.Wrap(err, "control plane reference to infrastructure machine not valid")
+	}
+	infra, err := r.getReference(ctx, cpInfraRef)
+	if err != nil {
+		return controlPlaneTopologyState{cp, nil}, errors.Wrap(err, "control plane infrastructure reference not found")
```

Comment:
WDYT, should we keep the error messages after getReference consistent?

Reply:
(I'm not sure the message is correct in all error cases of getReference)

Reply:
I'll make them more consistent. I'm really just trying to add the context of which component the error  is being encountered in.

Reply:
I think nearly everything is cleared up in the latest commit - PTAL when you have a chance.